game/src/game/player.rs

```diff
@@ -61,6 +132,563 @@ fn setup_player(q: Query<(&Transform, Entity), Added<PlayerBlueprint>>, mut comm
+    next: N,
+}
+
+impl<C: PlayerState, N: PlayerState> PlayerStateTransition<C, N> {
+    fn new(current: C, next: N) -> Self {
+        PlayerStateTransition { current, next }
+    }
+}
+
+pub trait Transitionable<T: PlayerState> {
+    type C: PlayerState + Default;
+
+    fn new_transition(&self, next: T) -> PlayerStateTransition<Self::C, T>
+    where
+        Self: PlayerState,
+    {
+        PlayerStateTransition::<Self::C, T>::new(Self::C::default(), next)
+    }
+}
+
+impl<T: PlayerState, N: PlayerState> Transitionable<N> for T
+where
+    T: PlayerState + Default,
+{
+    type C = Self;
+
+    fn new_transition(&self, next: N) -> PlayerStateTransition<Self::C, N>
+    where
+        Self: PlayerState,
+    {
+        PlayerStateTransition::<Self::C, N>::new(self.clone(), next)
+    }
+}
+
+fn transition<T: PlayerState, N: PlayerState>(
+    query: Query<(Entity, &PlayerStateTransition<T, N>)>,
+    mut commands: Commands,
+) {
+    for (entity, transition) in query.iter() {
+        commands
+            .entity(entity)
+            .insert(transition.next.clone())
+            .remove::<T>()
+            .remove::<PlayerStateTransition<T, N>>();
+    }
+}
+
+//======================================================================
+pub trait PlayerState: Component<Storage = SparseStorage> + Clone {}
+//could make derive macro to derive it?
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Idle;
+impl PlayerState for Idle {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Running;
+impl PlayerState for Running {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Falling;
+impl PlayerState for Falling {}
+
+#[derive(Component, Default, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Jumping {
+    //TODO:
+    //can this be frames/ticks instead?
+    airtime: Timer,
+}
+impl Jumping {
+    fn new() -> Self {
+        Jumping {
+            airtime: Timer::from_seconds(0.5, TimerMode::Once),
+        }
+    }
+}
+impl PlayerState for Jumping {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Grounded;
+impl PlayerState for Grounded {}
+
+#[derive(Component, Default, Copy, Clone)]
+#[component(storage = "SparseSet")]
+pub struct Attacking;
+impl PlayerState for Attacking {}
+
+//=======================================================================
+//
+//TODO: refactor, controller plugin, physics plugin, state plugin
+
+//maybe get rid of idle state?? could jsut play the animation when conditions are met. its a
+//composite of grounded + nothing else...
+fn player_add_idle(
+    query: Query<
+        (Entity, &ActionState<PlayerAction>),
+        (
+            With<PlayerGent>,
+            Without<Idle>,
+            Without<Running>,
+            Without<Jumping>,
+            Without<Falling>,
+            Without<Attacking>,
```

Comment:
This will quickly get unwieldy don't you think? And easy to forget to update as you add more new states in the future.

Reply:
I think so yea, this system is not actually being run/added to the schedule at all right now and I dont want to have to do something like this. I wrote it out to see if i felt like it made more sense as a way of dealing with idle but I don't like it for the same reasons you stated.